### ISSUE TITLE
Extension to method how zero-page section can be specified

### ISSUE THREAD

**Issue by laoo** [johnwbyrd/llvm-mos#6]
As a follow-up of a forum duscussion I've prepared a minimal proposal for zero-page section flag.

pros:

    it works
    it permits to have multiple different sections on page zero with different semantics
    it's not a precedence. There is already for example flag SHF_ARM_PURECODE with flag letter "y"

cons:
LLVM architecture is poorly suited to such actions and it must be done globally. E.g. the 'z' flag that I've used will work for any other machine as ElfArmParser does not check triples and does not forward unknown flags to target parsers. And vice versa - other flags for other machines can be interpreted as SHF_MOS_ZEROPAGE in MOSBackend if it happens to use the same bit. It can be easily done better in this place but in other places flags are treated globally too and handling it in implementations of specific machines needs heavy lifting.

**Comment by laoo** [comment]
I've redone the test-case hopefully better:

https://github.com/johnwbyrd/llvm-mos/blob/f329537f009a768732d11a421ce6dbf0cf8a629b/llvm/test/MC/MOS/zeropage.s#L1-L31

I've spotted issues though: the output of `llvm-mc -assemble --show-encoding` for the test is not what I've expected:
```	.text
.set adrAbsZ, 234
.set adrAbsA, 60138


adrNearA:

	lda	$ea                             ; encoding: [0xa5,0xea]
	lda	$eaea                           ; encoding: [0xad,0xea,0xea]
	lda	adrNearA                        ; encoding: [0xa5,A]
                                        ;   fixup A - offset: 1, value: adrNearA, kind: Addr8
	lda	adrFarZ1                        ; encoding: [0xa5,A]
                                        ;   fixup A - offset: 1, value: adrFarZ1, kind: Addr8
	lda	adrFarZ2                        ; encoding: [0xa5,A]
                                        ;   fixup A - offset: 1, value: adrFarZ2, kind: Addr8
	lda	adrFarA                         ; encoding: [0xa5,A]
                                        ;   fixup A - offset: 1, value: adrFarA, kind: Addr8

	.section	.zp,"",@nobits

adrFarZ1:
	.zero	1

	.section	customzp,"z",@nobits

adrFarZ2:
	.zero	1

	.section	nzp,"",@nobits

adrFarA:
	.zero	1
```

As you can see none of fixups are Addr16 even though resulting elf file is correct as the `objdump -d` shows:

```zeropage.o:	file format elf32-mos


Disassembly of section .text:

00000000 <adrNearA>:
       0: a5 ea        	lda	234
       2: ad ea ea     	lda	60138
       5: ad 00 00     	lda	0
       8: a5 00        	lda	0
       a: a5 00        	lda	0
       c: ad 00 00     	lda	0  
```

**Comment by johnwbyrd** [comment]
So what I think is happening here, is that running llvm-mc with the -assemble flag, outputs the result to an .s file.  And that .s file does not have relaxation logic applied to it.  That would be the point at which address lengths are determined for your code.

It seems that LLVM only wants to apply relaxation logic, including the code you added, only if the output is an .obj file.

So, in order to test this, the test will have to run something like the following commands:

; RUN: llvm-mc -triple mos --filetype=obj -o=%t.obj %s 
; RUN: llvm-objdump --all-headers --print-imm-hex -D %t.obj

And it will need to check the result against the known good outputs as you describe them above.  So in short, I think you are on the right track here -- one of us just needs to finish the test case to actually check the output of llvm-objdump, and reject the failing case.  I can do this when I get around to it -- you've done most of the hard work -- or you can.

**Comment by laoo** [comment]
OK. I understand now what happened and why the relaxation did not happen after running llvm-mc with the -assemble flag.
I'm still learning the LLVM infrastructure and I haven't figured out yet how automatic tests work.

**Comment by johnwbyrd** [comment]
Here is a high level overview of lit testing:

https://llvm.org/devmtg/2019-10/slides/Homerding-Kruse-LLVMTestingInfrastructureTutorial.pdf

Here is a mid level overview:

https://llvm.org/docs/TestSuiteGuide.html

Here is the documentation:

https://llvm.org/docs/CommandGuide/lit.html

**Comment by johnwbyrd** [comment]
Oh and by the way, please see my hotfix/laoo-zp-section branch for the current state of the fix to this bug.

**Comment by johnwbyrd** [comment]
Fixed this issue with 979f7da.
